crates/viewer/re_space_view_graph/src/graph/mod.rs

```diff
@@ -1,87 +1,185 @@
 mod hash;
+use std::hash::{Hash as _, Hasher as _};
+
+use egui::{Pos2, Vec2};
 pub(crate) use hash::GraphNodeHash;
 mod index;
-pub(crate) use index::NodeIndex;
+pub(crate) use index::NodeId;
 
-use re_types::components::{GraphNode, GraphType};
+use re_chunk::EntityPath;
+use re_log_types::Instance;
+use re_types::{components::GraphType, ArrowString};
 
-use crate::visualizers::{EdgeData, EdgeInstance, NodeData, NodeInstance};
+use crate::{
+    ui::DrawableLabel,
+    visualizers::{EdgeData, NodeData},
+};
 
-pub struct NodeInstanceImplicit {
-    pub node: GraphNode,
-    pub index: NodeIndex,
+pub enum Node {
+    Explicit {
+        id: NodeId,
+        instance: Instance,
+        node: ArrowString,
+        position: Option<Pos2>,
+        label: DrawableLabel,
+    },
+    Implicit {
```

Comment:
I've added doc comments, maybe this makes it clearer. Because `Node::Implicit` only has a very limited subset of information, I chose to introduce the enum at the top level to avoid any confusion. This is also the reason wy I opted against an `enum NodeCategory { Explicit, Implicit }`. 

We could also introduce a mixture of both approaches, but that would introduce another level of indirection, which I think can be avoided by a little redundancy.